### PR TITLE
Implementação dos testes e funcionalidade de Importar currículos de pesquisadores

### PR DESCRIPTION
**Pull request by PassaroDeFerro**
This change is [Reviewable](https://reviewable.io/reviews/pauloborba/rpa/11)

### CODE REVIEW

**Comment by pauloborba** [comment]
Reviewed 13 of 16 files at r2, 2 of 3 files at r4, 28 of 33 files at r5.
Review status: 43 of 49 files reviewed at latest revision, 7 unresolved discussions.

---

_[grails-app/controllers/rpa/ResearcherController.groovy, line 51 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWsp3M1CpS1i5sea-J3:-KWsp3M1CpS1i5sea-J4:b-ulfabs) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/grails-app/controllers/rpa/ResearcherController.groovy#L51)):_

> ``` Groovy
>                 researcherSaved = update()
>             } else {
>                 params << [name: researcherXml.name, cpf: researcherXml.cpf, articles: researcherXml.articles]
> ```

esse código foi copiado de outra contribuição?

---

_[grails-app/i18n/messages_it.properties, line 56 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWsmzdrLlQ9q9CfF92j:-KWsmzdrLlQ9q9CfF92k:br5ozfx) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/grails-app/i18n/messages_it.properties#L56)):_

> ``` INI
> typeMismatch.java.math.BigDecimal=La proprietà {0} deve essere un numero valido
> typeMismatch.java.math.BigInteger=La proprietà {0} deve essere un numero valido
> researcher.saved=The researcher below was successfully registered
> ```

não precisava ter alterado todos os arquivos de propriedade, nem misturado ingles com italiano

---

_[grails-app/services/rpa/XmlExtractorService.groovy, line 23 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWspHTq98nzmjWsjW2K:-KWspHTq98nzmjWsjW2L:b3jm7a0) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/grails-app/services/rpa/XmlExtractorService.groovy#L23)):_

> ``` Groovy
>                 for (int i = 0; i < nodeList.getLength(); i++) {
>                     Node node = nodeList.item(i);
>                     if (node instanceof Element) {
> ```

refactor? copiado de outra contribuição?

---

_[test/cucumber/steps/ImportarLattesSteps.groovy, line 17 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWsnuFiUHAfs8RgX07h:-KWsnuFiUHAfs8RgX07i:bdmv3ta) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/test/cucumber/steps/ImportarLattesSteps.groovy#L17)):_

> ``` Groovy
>     assert (pesquisador.cpf == cpf) && (pesquisador.name == nome)
> }
> //Funciona :)
> ```

tira esses comentários

---

_[test/functional/pages/CreateResearcherPage.groovy, line 10 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWsoGIR2FgluYO515bh:-KWsoGIR2FgluYO515bi:blk4ugs) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/test/functional/pages/CreateResearcherPage.groovy#L10)):_

> ``` Groovy
> 
>     static at = {
>         title ==~ /Criar Pesquisador/ || title ==~ /Create Researcher/
> ```

internacionalizar

---

_[test/functional/pages/ShowResearcherPage.groovy, line 13 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWso_OBmq3JHGoClQTa:-KWso_OBmq3JHGoClQTb:b-m7u6mi) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/test/functional/pages/ShowResearcherPage.groovy#L13)):_

> ``` Groovy
> 
>     def findAcceptedMsg(){
>         assert $("div", class: "message").text() == "O pesquisador abaixo foi registrado com sucesso" ||
> ```

internacionaliza

---

_[test/functional/steps/TestAndOperations.groovy, line 21 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWsooFThLgPlVblLjZ3:-KWsooFUD4OlcFZ8NfOX:blc3v82) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/test/functional/steps/TestAndOperations.groovy#L21)):_

> ``` Groovy
>     }
> 
>     static public Researcher buildResearcher(String name, String cpf){
> ```

para que isso? nao deveria ser o createReasearcher mais abaixo?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/11)_

**Comment by pauloborba** [comment]
Review status: 38 of 49 files reviewed at latest revision, 8 unresolved discussions.

---

*[test/cucumber/ImportarLattes.feature, line 6 at r6](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KX5qkI2ZGLoaNuGA7I0:-KX5qkI2ZGLoaNuGA7I1:b88lh3d) ([raw file](https://github.com/pauloborba/rpa/blob/af3d64cfe5a9020f2f803f80292f46628885f9d0/test/cucumber/ImportarLattes.feature#L6)):*
> ```Cucumber
>   So that eu possa visualiza-los e compara-los depois
> 
> #Controller
> ```

cenários de falha?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/11)*

File: .travis.yml [hunk 1]
```diff
@@ -18,4 +18,4 @@ branches:
   only:
     - master
 
-script: grails -Dgeb.env=firefox test-app --non-interactive --stacktrace
```
Comment (rafaellsma) [inline .travis.yml:21]: Nao entendi por qual motivo você adicionou isso.

File: grails-app/controllers/rpa/ResearcherController.groovy [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        if (!researcherInstance) {
+            flash.message = message(code: 'default.not.found.message', args: [message(code: 'researcher.label', default: 'Pesquisador'), params.id])
+            return
+        }
+        [researcherInstance: researcherInstance]
+    }
+
+    def save(){
+        Researcher r = new Researcher(params)
+        r.articles.each {
```
Comment (rafaellsma) [inline grails-app/controllers/rpa/ResearcherController.groovy:25]: Poderíamos refatorar essa parte, colocando-a no domain.

File: grails-app/controllers/rpa/ResearcherController.groovy [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    def update(){
+        Researcher researcherSaved = params['researcherSaved']
+        Researcher researcherNew = params['researcherNew']
+        researcherSaved.save flush:true
+    }
+
+    def importFile(){
+        def xml = request.getFile('file')
+        if(!xml.empty){
+            XmlExtractorService xmlExtractor = new XmlExtractorService()
```
Comment (rafaellsma) [inline grails-app/controllers/rpa/ResearcherController.groovy:34]: No lugar de chamarmos duas novas action, poderíamos colocar tudo apenas nessa.

File: grails-app/i18n/messages_cs_CZ.properties [hunk 4]
```diff
@@ -53,3 +53,4 @@ typeMismatch.java.lang.Long=Položka {0} musí být validní číslo
 typeMismatch.java.lang.Short=Položka {0} musí být validní číslo
 typeMismatch.java.math.BigDecimal=Položka {0} musí být validní číslo
 typeMismatch.java.math.BigInteger=Položka {0} musí být validní číslo
+researcher.saved=The researcher below was successfully registered
```
Comment (rafaellsma) [inline grails-app/i18n/messages_cs_CZ.properties]: Tu adicionou um mensagem em inglês nas messages que deveriam está em tcheco. Poderia botar em tcheco ou não colocar, e caso o sistema tivesse em tcheco usaria a message default. Esse mesmo erro acontece em outras línguas

File: grails-app/services/rpa/XmlExtractorService.groovy [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+class XmlExtractorService {
+    def getResearcher(InputStream file){
+        Researcher researcher = new Researcher()
+        try {
+            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
```
Comment (rafaellsma) [inline grails-app/services/rpa/XmlExtractorService.groovy]: Temos que refatorar(extract method) melhor

File: test/cucumber/ImportarLattes.feature [hunk 6]
```diff
@@ -0,0 +1,50 @@
+Feature: Importar curriculos lattes
+  As um usuário
+  I want to importar curriculos lattes
+  So that eu possa visualiza-los e compara-los depois
+
+#Controller
+Scenario: Cadastrar um pesquisador com vários artigos com sucesso
+Given O arquivo "CurriculoValidoComDoisArtigos.xml" contém um pesquisador de CPF "12345678911" e nome "Rafael".
+And   O arquivo "CurriculoValidoComDoisArtigos.xml" contém o artigo "Implementation cow confident with BFS".
+And   O arquivo "CurriculoValidoComDoisArtigos.xml" contém o artigo "Implementation sudoku with backtracking".
```
Comment (rafaellsma) [inline test/cucumber/ImportarLattes.feature:11]: Poderia especificar melhor que o arquivo só tem isso e nada mais

File: test/cucumber/ImportarLattes.feature [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+And   O pesquisador de CPF "12345678911" não está cadastrado
+When  Um arquivo de currículo de nome "CurriculoValidoComDoisArtigos.xml" é importado
+Then  O pesquisador de nome "Rafael" e CPF "12345678911" é cadastrado no sistema
+And   O artigo "Implementation cow confident with BFS" é adicionado ao currículo do pesquisador de CPF "12345678911".
+And   O artigo "Implementation sudoku with backtracking" é adicionado ao currículo do pesquisador de CPF "12345678911".
+
+#GUI
+Scenario: Cadastrar um pesquisador com vários artigos com sucesso
+Given Estou na página de cadastrar pesquisadores
+And   O arquivo "CurriculoValidoComDoisArtigos.xml" contém um pesquisador de CPF "12345678911" e nome "Rafael".
```
Comment (rafaellsma) [inline test/cucumber/ImportarLattes.feature]: Alguns passos de GUI são so mesmos de controller, isso mistura controller com gui.

File: test/cucumber/ImportarLattes.feature [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+When  Um arquivo de currículo de nome "CurriculoValidoSemArtigos.xml" é importado
+Then  O pesquisador de nome "Fulano" e CPF "01234567890" é cadastrado no sistema
+And   O pesquisador de CPF "01234567890" não possui artigos
+
+#GUI
+Scenario: Cadastrar um pesquisador sem artigos com sucesso
+Given Estou na página de cadastrar pesquisadores
+And   O arquivo "CurriculoValidoSemArtigos.xml" contém um pesquisador de CPF "01234567890" e nome "Fulano".
+And   O arquivo "CurriculoValidoSemArtigos.xml" não contem artigos
+And   O pesquisador de CPF "01234567890" não está cadastrado
```
Comment (rafaellsma) [inline test/cucumber/ImportarLattes.feature]: novamento misturando gui e controller

File: test/cucumber/steps/ImportarLattesSteps.groovy [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import cucumber.api.PendingException
+
+import static cucumber.api.groovy.EN.*
+import static steps.TestAndOperations.*
+import rpa.Researcher
+import pages.CreateResearcherPage
+import pages.ShowResearcherPage
+
+//---------- GIVENs ----------
+//Funciona :)
```
Comment (rafaellsma) [inline test/cucumber/steps/ImportarLattesSteps.groovy]: Poderia remover esse comentários ja que não são relevantes.

File: test/cucumber/steps/ImportarLattesSteps.groovy [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+Given(~/^O arquivo "([^"]*)" contém um pesquisador de CPF "([^"]*)" e nome "([^"]*)"\.$/) { String arquivo, String cpf, String nome ->
+    def pesquisador = buildResearcherWithFile(arquivo)
+    assert (pesquisador.cpf == cpf) && (pesquisador.name == nome)
+}
+//Funciona :)
+And(~/^O pesquisador de CPF "([^"]*)" não está cadastrado$/) { String cpf ->
```
Comment (rafaellsma) [inline test/cucumber/steps/ImportarLattesSteps.groovy:15]: Esse step e outros estão sendo chamados por GUI e Controller, sendo executado códigos pertinentes somente ao teste de controller.

File: test/functional/pages/CreateResearcherPage.groovy [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package pages
+
+import geb.Page
+
+class CreateResearcherPage extends Page {
+    private static final dirFiles = "test/functional/steps/"
+    static url = "/RPA/researcher/create/"
+
+    static at = {
+        title ==~ /Create Researcher/
```
Comment (rafaellsma) [inline test/functional/pages/CreateResearcherPage.groovy]: Falta internacionalizar esses testes.

**Comment by pauloborba** [comment]
Reviewed 13 of 26 files at r7.
Review status: 34 of 44 files reviewed at latest revision, 19 unresolved discussions.

---

*[grails-app/controllers/rpa/ResearcherController.groovy, line 25 at r7](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KY-qrSSqG2w2SKQHGKv:-KY-qrSSqG2w2SKQHGKw:b-czh41z) ([raw file](https://github.com/pauloborba/rpa/blob/4b188818e70e8d3d82d7c3611853d0602cc68947/grails-app/controllers/rpa/ResearcherController.groovy#L25)):*
> ```Groovy
>     def save(){
>         Researcher r = new Researcher(params)
>         r.articles.each {
> ```

não parece fazer sentido

---

*[grails-app/controllers/rpa/ResearcherController.groovy, line 32 at r7](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KY-qosuOqjBb8j25VZu:-KY-qosuOqjBb8j25VZv:b-jcffjl) ([raw file](https://github.com/pauloborba/rpa/blob/4b188818e70e8d3d82d7c3611853d0602cc68947/grails-app/controllers/rpa/ResearcherController.groovy#L32)):*
> ```Groovy
> 
>     def update(){
> ```

vazio?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/11)*

File: grails-app/domain/rpa/Article.groovy [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package rpa
+
+class Article {
+    String tittle
+    String journal
+    String issn
+    static belongsTo = [owner:Researcher]
```
Comment (rafaellsma) [inline grails-app/domain/rpa/Article.groovy:7]: Faltou adicionar os autores

**Comment by PassaroDeFerro** [comment]
Review status: 34 of 45 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[grails-app/i18n/messages_it.properties, line 56 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWsmzdrLlQ9q9CfF92j:-KYxbcVVQ9gagfM9Shg2:byuvsxz) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/grails-app/i18n/messages_it.properties#L56)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

não precisava ter alterado todos os arquivos de propriedade, nem misturado ingles com italiano
</blockquote></details>

Aconteceu porque eu gerei a propriedade, isso já foi corrigido.

---

*[test/cucumber/ImportarLattes.feature, line 6 at r6](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KX5qkI2ZGLoaNuGA7I0:-KYxcXJosjhepjbydtHl:b-nmwry8) ([raw file](https://github.com/pauloborba/rpa/blob/af3d64cfe5a9020f2f803f80292f46628885f9d0/test/cucumber/ImportarLattes.feature#L6)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

cenários de falha?
</blockquote></details>

Quatro cenários (2 de GUI e 2 de Controller) feitos

---

*[test/cucumber/ImportarLattes.feature, line 20 at r6](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWFXNwJ9lTOezlPCtI--r6-20:-KYxc4lTsf2qlrmDu6iU:bknwqmw) ([raw file](https://github.com/pauloborba/rpa/blob/af3d64cfe5a9020f2f803f80292f46628885f9d0/test/cucumber/ImportarLattes.feature#L20)):*
<details><summary><i>Previously, rafaellsma (Rafael Albuquerque) wrote…</i></summary><blockquote>

Alguns passos de GUI são so mesmos de controller, isso mistura controller com gui.
</blockquote></details>

Os passos que checam conteúdo de arquivos foram usados em ambos porque o que a GUI exibe precisa estar em conformidade com o que tinha no arquivo

---

*[test/cucumber/steps/ImportarLattesSteps.groovy, line 17 at r5](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWsnuFiUHAfs8RgX07h:-KYxcmxFVgCsjd4_3Izb:b7sazpu) ([raw file](https://github.com/pauloborba/rpa/blob/5d233f1bdde6589238523bfecb00ce5caff673de/test/cucumber/steps/ImportarLattesSteps.groovy#L17)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

tira esses comentários
</blockquote></details>

Eles já foram removidos

---

*[test/cucumber/steps/ImportarLattesSteps.groovy, line 18 at r6](https://reviewable.io:443/reviews/pauloborba/rpa/11#-KWFXNwJ9lTOezlPCtI0-r6-18:-KYxcsAjLi038rbUZTbz:b5fku15) ([raw file](https://github.com/pauloborba/rpa/blob/af3d64cfe5a9020f2f803f80292f46628885f9d0/test/cucumber/steps/ImportarLattesSteps.groovy#L18)):*
<details><summary><i>Previously, rafaellsma (Rafael Albuquerque) wrote…</i></summary><blockquote>

Esse step e outros estão sendo chamados por GUI e Controller, sendo executado códigos pertinentes somente ao teste de controller.
</blockquote></details>

Agora só os testes de Controller usam esse passo, os de GUI usam outro

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/11)*

File: test/cucumber/ImportarLattes.feature (hunk 6 above)
Comment (Alexsandro-Vitor) [inline test/cucumber/ImportarLattes.feature]: Agora existe um passo que diz quantos artigos o pesquisador tem

File: grails-app/services/rpa/XmlExtractorService.groovy (hunk 5 above)
Comment (PassaroDeFerro) [reply grails-app/services/rpa/XmlExtractorService.groovy]: O XmlExtractor foi refatorado

**Pull request closed without merging**